File: Plugins/aiproviders/flashplugin.py
"""Minimal Flash plugin protocol shim for Python (stdlib only).

Speaks the wire contract from docs/plugin-protocol.md: length-prefixed
MessagePack over stdio (4-byte big-endian length + one value), protocol v3
lifecycle (initialize/heartbeat/shutdown), and request dispatch. Hand-rolls
the MessagePack subset the protocol needs — nil/bool/int/str/array/map —
so a plugin author needs nothing beyond /usr/bin/python3.
"""
import os
import struct
import sys


def encode(obj):
    if obj is None:
        return b"\xc0"
    if obj is True:
        return b"\xc3"
    if obj is False:
        return b"\xc2"
    if isinstance(obj, int):
        if 0 <= obj <= 127:
            return struct.pack("B", obj)
        if -32 <= obj < 0:
            return struct.pack("b", obj)
        return b"\xd3" + struct.pack(">q", obj)
    if isinstance(obj, str):
        raw = obj.encode()
        if len(raw) < 32:
            return struct.pack("B", 0xA0 | len(raw)) + raw
        return b"\xdb" + struct.pack(">I", len(raw)) + raw
    if isinstance(obj, (list, tuple)):
        head = (
            struct.pack("B", 0x90 | len(obj))
            if len(obj) < 16
            else b"\xdc" + struct.pack(">H", len(obj))
        )
        return head + b"".join(encode(x) for x in obj)
    if isinstance(obj, dict):
        head = (
            struct.pack("B", 0x80 | len(obj))
            if len(obj) < 16
            else b"\xde" + struct.pack(">H", len(obj))
        )
        return head + b"".join(encode(k) + encode(v) for k, v in obj.items())
    raise TypeError(f"unencodable: {type(obj)}")


def decode(buf, pos=0):
    b = buf[pos]
    pos += 1
    if b == 0xC0:
        return None, pos
    if b == 0xC2:
        return False, pos
    if b == 0xC3:
        return True, pos
    if b <= 0x7F:
        return b, pos
    if b >= 0xE0:
        return b - 256, pos
    if 0xA0 <= b <= 0xBF:
        n = b & 0x1F
        return buf[pos : pos + n].decode("utf-8", "replace"), pos + n
    if b == 0xD9:
        n = buf[pos]
        return buf[pos + 1 : pos + 1 + n].decode("utf-8", "replace"), pos + 1 + n
    if b == 0xDA:
        (n,) = struct.unpack(">H", buf[pos : pos + 2])
        return buf[pos + 2 : pos + 2 + n].decode("utf-8", "replace"), pos + 2 + n
    if b == 0xDB:
        (n,) = struct.unpack(">I", buf[pos : pos + 4])
        return buf[pos + 4 : pos + 4 + n].decode("utf-8", "replace"), pos + 4 + n
    if 0x80 <= b <= 0x8F or b in (0xDE, 0xDF):
        if b <= 0x8F:
            n = b & 0x0F
        elif b == 0xDE:
            (n,) = struct.unpack(">H", buf[pos : pos + 2])
            pos += 2
        else:
            (n,) = struct.unpack(">I", buf[pos : pos + 4])
            pos += 4
        out = {}
        for _ in range(n):
            k, pos = decode(buf, pos)
            v, pos = decode(buf, pos)
            out[k] = v
        return out, pos
    if 0x90 <= b <= 0x9F or b in (0xDC, 0xDD):
        if b <= 0x9F:
            n = b & 0x0F
        elif b == 0xDC:
            (n,) = struct.unpack(">H", buf[pos : pos + 2])
            pos += 2
        else:
            (n,) = struct.unpack(">I", buf[pos : pos + 4])
            pos += 4
        out = []
        for _ in range(n):
            v, pos = decode(buf, pos)
            out.append(v)
        return out, pos
    if b in (0xCC, 0xD0):
        return buf[pos], pos + 1
    if b in (0xCD, 0xD1):
        return struct.unpack(">H", buf[pos : pos + 2])[0], pos + 2
    if b in (0xCE, 0xD2):
        return struct.unpack(">I", buf[pos : pos + 4])[0], pos + 4
    if b in (0xCF, 0xD3):
        return struct.unpack(">q", buf[pos : pos + 8])[0], pos + 8
    if b == 0xCA:
        return struct.unpack(">f", buf[pos : pos + 4])[0], pos + 4
    if b == 0xCB:
        return struct.unpack(">d", buf[pos : pos + 8])[0], pos + 8
    raise ValueError(f"unhandled msgpack byte 0x{b:02x}")


PROTOCOL_VERSION = 3


class Plugin:
    """Blocking single-threaded serve loop. Enough for command-style
    plugins; a sources plugin would publish its warm catalog before
    replying to initialize (see the readiness gate in the protocol doc)."""

    def __init__(self, published_sources=()):
        self.published_sources = list(published_sources)
        self._out = sys.stdout.buffer

    def send(self, obj):
        payload = encode(obj)
        self._out.write(struct.pack(">I", len(payload)) + payload)
        self._out.flush()

    def respond(self, request_id, result):
        self.send({"jsonrpc": "2.0", "id": request_id, "result": result})

    def log(self, level, message):
        self.send(
            {
                "jsonrpc": "2.0",
                "method": "flash.log",
                "params": {"level": level, "message": message, "fields": {}},
            }
        )

    def serve(self, on_command):
        stdin = sys.stdin.buffer
        while True:
            header = stdin.read(4)
            if len(header) < 4:
                return  # host closed stdin — exit
            (n,) = struct.unpack(">I", header)
            payload = stdin.read(n)
            msg, _ = decode(payload)
            method = msg.get("method")
            request_id = msg.get("id")
            if method == "initialize":
                version = msg.get("params", {}).get("protocol_version")
                if version != PROTOCOL_VERSION:
                    self.respond(
                        request_id,
                        {"ok": False, "error": f"protocol {version} != {PROTOCOL_VERSION}"},
                    )
                    return
                self.respond(
                    request_id,
                    {
                        "ok": True,
                        "protocol_version": PROTOCOL_VERSION,
                        "published_sources": self.published_sources,
                    },
                )
            elif method == "heartbeat":
                self.respond(request_id, {"ok": True})
            elif method == "shutdown":
                self.respond(request_id, {"ok": True})
                return
            elif method == "command.invoke":
                self.respond(request_id, on_command(msg.get("params", {})))
            elif request_id is not None:
                self.respond(
                    request_id, {"ok": False, "error": f"unsupported method {method}"}
                )


def data_dir():
    return os.environ.get("FLASH_PLUGIN_DATA_DIR", ".")

File: Plugins/emojis/flashplugin.ts
// Minimal Flash plugin protocol shim for TypeScript on Bun.
//
// Speaks the wire contract from docs/plugin-protocol.md: length-prefixed
// MessagePack over stdio (4-byte big-endian length + one value), protocol v3
// lifecycle, and the warm-catalog readiness gate — a sources plugin must
// publish its canonical `plugin:<id>` catalog before initialize succeeds.
// Hand-rolls the MessagePack subset the protocol needs, so a plugin author
// needs nothing beyond the Bun runtime.

export type Value =
  | null
  | boolean
  | number
  | string
  | Value[]
  | { [key: string]: Value };

export function encode(obj: Value): Uint8Array {
  const parts: Uint8Array[] = [];
  encodeInto(obj, parts);
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let at = 0;
  for (const p of parts) {
    out.set(p, at);
    at += p.length;
  }
  return out;
}

function u8(...bytes: number[]): Uint8Array {
  return new Uint8Array(bytes);
}

function be32(n: number): Uint8Array {
  const out = new Uint8Array(4);
  new DataView(out.buffer).setUint32(0, n);
  return out;
}

function encodeInto(obj: Value, parts: Uint8Array[]) {
  if (obj === null) return void parts.push(u8(0xc0));
  if (obj === true) return void parts.push(u8(0xc3));
  if (obj === false) return void parts.push(u8(0xc2));
  if (typeof obj === "number") {
    if (Number.isInteger(obj) && obj >= 0 && obj <= 127)
      return void parts.push(u8(obj));
    if (Number.isInteger(obj) && obj >= -32 && obj < 0)
      return void parts.push(u8(obj & 0xff));
    if (Number.isInteger(obj)) {
      const out = new Uint8Array(9);
      out[0] = 0xd3;
      new DataView(out.buffer).setBigInt64(1, BigInt(obj));
      return void parts.push(out);
    }
    const out = new Uint8Array(9);
    out[0] = 0xcb;
    new DataView(out.buffer).setFloat64(1, obj);
    return void parts.push(out);
  }
  if (typeof obj === "string") {
    const raw = new TextEncoder().encode(obj);
    if (raw.length < 32) parts.push(u8(0xa0 | raw.length));
    else {
      parts.push(u8(0xdb));
      parts.push(be32(raw.length));
    }
    return void parts.push(raw);
  }
  if (Array.isArray(obj)) {
    if (obj.length < 16) parts.push(u8(0x90 | obj.length));
    else {
      parts.push(u8(0xdc, obj.length >> 8, obj.length & 0xff));
    }
    for (const x of obj) encodeInto(x, parts);
    return;
  }
  const keys = Object.keys(obj);
  if (keys.length < 16) parts.push(u8(0x80 | keys.length));
  else parts.push(u8(0xde, keys.length >> 8, keys.length & 0xff));
  for (const k of keys) {
    encodeInto(k, parts);
    encodeInto(obj[k], parts);
  }
}

export function decode(buf: Uint8Array, pos = 0): [Value, number] {
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  const b = buf[pos++];
  if (b === 0xc0) return [null, pos];
  if (b === 0xc2) return [false, pos];
  if (b === 0xc3) return [true, pos];
  if (b <= 0x7f) return [b, pos];
  if (b >= 0xe0) return [b - 256, pos];
  const str = (n: number) => {
    const s = new TextDecoder().decode(buf.subarray(pos, pos + n));
    return [s, pos + n] as [Value, number];
  };
  if (b >= 0xa0 && b <= 0xbf) return str(b & 0x1f);
  if (b === 0xd9) return ((pos += 1), str(buf[pos - 1]));
  if (b === 0xda) return ((pos += 2), str(view.getUint16(pos - 2)));
  if (b === 0xdb) return ((pos += 4), str(view.getUint32(pos - 4)));
  const container = (n: number, isMap: boolean): [Value, number] => {
    if (isMap) {
      const out: { [k: string]: Value } = {};
      for (let i = 0; i < n; i++) {
        const [k, p1] = decode(buf, pos);
        const [v, p2] = decode(buf, p1);
        out[String(k)] = v;
        pos = p2;
      }
      return [out, pos];
    }
    const out: Value[] = [];
    for (let i = 0; i < n; i++) {
      const [v, p] = decode(buf, pos);
      out.push(v);
      pos = p;
    }
    return [out, pos];
  };
  if (b >= 0x80 && b <= 0x8f) return container(b & 0x0f, true);
  if (b >= 0x90 && b <= 0x9f) return container(b & 0x0f, false);
  if (b === 0xde) return ((pos += 2), container(view.getUint16(pos - 2), true));
  if (b === 0xdf) return ((pos += 4), container(view.getUint32(pos - 4), true));
  if (b === 0xdc) return ((pos += 2), container(view.getUint16(pos - 2), false));
  if (b === 0xdd) return ((pos += 4), container(view.getUint32(pos - 4), false));
  if (b === 0xcc || b === 0xd0) return [buf[pos], pos + 1];
  if (b === 0xcd || b === 0xd1) return [view.getUint16(pos), pos + 2];
  if (b === 0xce || b === 0xd2) return [view.getUint32(pos), pos + 4];
  if (b === 0xcf || b === 0xd3)
    return [Number(view.getBigUint64(pos)), pos + 8];
  if (b === 0xca) return [view.getFloat32(pos), pos + 4];
  if (b === 0xcb) return [view.getFloat64(pos), pos + 8];
  throw new Error(`unhandled msgpack byte 0x${b.toString(16)}`);
}

export const PROTOCOL_VERSION = 3;

export interface Candidate {
  title: string;
  url?: string;
  metadata: { [key: string]: string };
}

type Msg = { [key: string]: Value };

export class Plugin {
  private warm = new Map<string, Candidate[]>();
  // One writer for the process lifetime, flushed per frame: Bun's FileSink
  // buffers, and a fresh writer per send can lose the final (shutdown)
  // reply when the process exits before the sink drains.
  private writer = Bun.stdout.writer();

  setLocations(sourceId: string, candidates: Candidate[]) {
    this.warm.set(sourceId, candidates);
  }

  private send(obj: Value) {
    const payload = encode(obj);
    const framed = new Uint8Array(4 + payload.length);
    new DataView(framed.buffer).setUint32(0, payload.length);
    framed.set(payload, 4);
    this.writer.write(framed);
    this.writer.flush();
  }

  private respond(id: Value, result: Value) {
    this.send({ jsonrpc: "2.0", id, result });
  }

  log(level: string, message: string) {
    this.send({
      jsonrpc: "2.0",
      method: "flash.log",
      params: { level, message, fields: {} },
    });
  }

  async serve() {
    let buf = new Uint8Array(0);
    for await (const chunk of Bun.stdin.stream()) {
      const next = new Uint8Array(buf.length + chunk.length);
      next.set(buf);
      next.set(chunk, buf.length);
      buf = next;
      while (buf.length >= 4) {
        const n = new DataView(buf.buffer, buf.byteOffset).getUint32(0);
        if (buf.length < 4 + n) break;
        const [msg] = decode(buf.subarray(4, 4 + n));
        buf = buf.slice(4 + n);
        if (!this.dispatch(msg as Msg)) return;
      }
    }
  }

  private dispatch(msg: Msg): boolean {
    const method = msg.method as string | undefined;
    const id = msg.id as Value;
    switch (method) {
      case "initialize": {
        const params = (msg.params ?? {}) as Msg;
        if (params.protocol_version !== PROTOCOL_VERSION) {
          this.respond(id, {
            ok: false,
            error: `protocol ${params.protocol_version} != ${PROTOCOL_VERSION}`,
          });
          return false;
        }
        this.respond(id, {
          ok: true,
          protocol_version: PROTOCOL_VERSION,
          published_sources: [...this.warm.keys()].sort(),
        });
        return true;
      }
      case "heartbeat":
        this.respond(id, { ok: true });
        return true;
      case "shutdown":
        this.respond(id, { ok: true });
        return false;
      case "sources.snapshot": {
        const candidates = [...this.warm.entries()]
          .sort(([a], [b]) => a.localeCompare(b))
          .flatMap(([, rows]) => rows);
        this.respond(id, { candidates: candidates as unknown as Value });
        return true;
      }
      default:
        if (id !== undefined && id !== null) {
          this.respond(id, { ok: false, error: `unsupported method ${method}` });
        }
        return true;
    }
  }
}
